### ISSUE TITLE
How to get the optimal trajectory, the dynamics and the map ?

### ISSUE THREAD

**Issue by jrabary** [duckietown/gym-duckietown-agent#5]
According to the documentation the reward is based on how well the robot follows the optimal racing trajectory. Is it possible to have access to this optimal trajectory, the dynamics of the robot, the map used in the environment in order the create a controller based on these information ?

**Comment by fgolemo** [comment]
Hey there Jaonary,

Short answer: no, because then it would be too easy. :grinning:.

Long answer:
It depends on the task: for the `LF`/`LFV` tasks you don't have access to the map. For the `NAVV` task you do (because you have to do some kind of start-goal planning) and it will be included in the `misc` dictionary in each observation. We are currently implementing this task and it should be ready soon.

For the optimal trajectory: You can see it projected onto the road (as a means of debugging) if you launch the gym-server on its own. See in this file https://github.com/duckietown/gym-duckietown/blob/master/manual_control.py#L20 Line 20, the command line switch to activate the lane following trajectory overlay. However this is just for debugging. We are also considering forwarding this information to the `gym-duckietown-agent` so that you as the human (not the actual robot) can see the line. Your robot has to learn where this line is, by looking at the reward.

As for the dynamics: What exactly do you need?

Maybe @maximecb  can say something about how https://github.com/duckietown/gym-duckietown/blob/master/experiments/control_pid.py works?

**Comment by maximecb** [comment]
So, the important distinction here is, yes you can absolutely get information about the idealised lane-following curve while training. You can use the `experiments/control_pid.py` script to generate data which you then use to train an imitation learning agent, for example (look at the source to see how this works). You can also modify the simulator and access all of its internal parameters as you wish, for the sake of generating training data.

However, at evaluation time, during the competition, you won't have access to this information. Your code will have to run inside of a `gym-duckietown-agent` docker container, and you will only get a series of images as input. You could train a neural network to predict information about the curve, and use this to drive your agent, however.

**Comment by jrabary** [comment]
@maximecb @fgolemo Thanks for this answer. In fact, I would like to use a PID or a MPC controller to sample trajectories. The link you provided is a good starting point.
